traffic_ops/traffic_ops_golang/user/user.go

```diff
@@ -0,0 +1,445 @@
+func (user *TOUser) Create() (error, error, int) {
+
+	// PUT and POST validation differs slightly
+	err := user.postValidate()
+	if err != nil {
+		return err, nil, http.StatusBadRequest
+	}
+
+	// Convert password to SCRYPT
+	*user.LocalPassword, err = auth.DerivePassword(*user.LocalPassword)
+	if err != nil {
+		return err, nil, http.StatusBadRequest
+	}
+	*user.ConfirmLocalPassword, err = auth.DerivePassword(*user.ConfirmLocalPassword)
+	if err != nil {
+		return err, nil, http.StatusBadRequest
+	}
+
+	resultRows, err := user.ReqInfo.Tx.NamedQuery(user.InsertQuery(), user)
+	if err != nil {
+		return api.ParseDBError(err)
+	}
+	defer resultRows.Close()
+
+	var id int
+	var lastUpdated tc.TimeNoMod
+	var tenant string
+	var rolename string
+
+	rowsAffected := 0
+	for resultRows.Next() {
+		rowsAffected++
+		if err = resultRows.Scan(&id, &lastUpdated, &tenant, &rolename); err != nil {
+			return nil, fmt.Errorf("could not scan after insert: %s\n)", err), http.StatusInternalServerError
+		}
+	}
+
+	if rowsAffected == 0 {
+		return nil, fmt.Errorf("no user was inserted, nothing was returned"), http.StatusInternalServerError
+	} else if rowsAffected > 1 {
+		return nil, fmt.Errorf("too many rows affected from user insert"), http.StatusInternalServerError
+	}
+
+	user.ID = &id
+	user.LastUpdated = &lastUpdated
+	user.Tenant = &tenant
+	user.RoleName = &rolename
+	user.LocalPassword = nil
+	user.ConfirmLocalPassword = nil
+
+	return nil, nil, http.StatusOK
+}
+
+// returning true indicates the data related to the given tenantID should be visible
+// this is just a linear search;`tenantIDs` is presumed to be unsorted
+func checkTenancy(tenantID *int, tenantIDs []int) bool {
+	for _, id := range tenantIDs {
+		if id == *tenantID {
+			return true
+		}
+	}
+	return false
+}
+
+// This is not using GenericRead because of this tenancy check. Maybe we can add tenancy functionality to the generic case?
+func (user *TOUser) Read() ([]interface{}, error, error, int) {
+
+	tenantIDs, err := tenant.GetUserTenantIDListTx(user.ReqInfo.Tx.Tx, user.ReqInfo.User.TenantID)
+	if err != nil {
+		return nil, nil, fmt.Errorf("getting tenant list for user: %v\n", err), http.StatusInternalServerError
+	}
+
+	var query string
+	where, orderBy, queryValues, errs := dbhelpers.BuildWhereAndOrderBy(user.APIInfo().Params, user.ParamColumns())
+
+	if len(errs) > 0 {
+		return nil, util.JoinErrs(errs), nil, http.StatusBadRequest
+	}
+
+	query = user.SelectQuery() + where + orderBy
+	rows, err := user.ReqInfo.Tx.NamedQuery(query, queryValues)
+	if err != nil {
+		return nil, nil, fmt.Errorf("querying users : %v", err), http.StatusInternalServerError
+	}
+	defer rows.Close()
+
+	users := []interface{}{}
+	for rows.Next() {
+
+		if err = rows.StructScan(user); err != nil {
+			return nil, nil, fmt.Errorf("parsing user rows: %v", err), http.StatusInternalServerError
+		}
+
+		if checkTenancy(user.TenantID, tenantIDs) {
```

Comment:
instead of doing this for each row,   better to use `dbhelpers.AddTenancyCheck` to add to the sql query.  The rows you end up with will all be within allowed tenants.  See implementation of `DeliveryserviceV13.Read` for an example.